Library/Homebrew/cmd/link.rb

```diff
@@ -40,18 +42,34 @@ def link
       verbose:   args.verbose?,
     }
 
-    args.named.to_kegs.each do |keg|
+    kegs = if args.HEAD?
+      args.named.to_kegs.group_by(&:name).transform_values do |v|
+        v.find do |keg|
+          keg.version.head?
+        end
+      end
+    else
+      args.named.to_latest_kegs.group_by(&:name).transform_values(&:first)
```

Comment:
Doesn't `to_latest_kegs` already only return one per argument? I feel like we should be able to just use `args.named.to_latest_kegs` here without needing the `group_by` or `transform_values`. Am I missing something?

Reply:
> Doesn't to_latest_kegs already only return one per argument?

You're correct. `to_latest_kegs` works fine on its own without the `group_by` and `transform_values`. 
However, `to_kegs` causes trouble when you filter its return value for `head` kegs. There's no way to tell If there are no `head` kegs installed for one of the named arguments. `to_kegs` only throws an error when there are no kegs installed for a named argument.

```ruby
args.named.to_kegs.group_by(&:name).transform_values do |v|
  v.find do |keg|
    keg.version.head?
  end
end
```
The code above converts the array of kegs to a hash that maps formula name to an array of corresponding kegs. Then, it transforms the hash values to be the head keg if it exists or `nil` if not. This allows subsequent code to print a warning if `--HEAD` is given but a named argument doesn't have a corresponding `head` keg installed.

```ruby
args.named.to_latest_kegs.group_by(&:name).transform_values(&:first)
```
The code above works similarly except it uses `first` to transform all of the one-element array values to kegs. This is done so that the `kegs` hash has the same structure whether `--HEAD` is given or not.

I personally think that `to_kegs` shouldn't flat map all kegs corresponding to named arguments into a single array, but I didn't want to modify all of the code that depends on `to_kegs` for such an edge case.

Reply:
Yeah, I agree with the `--HEAD` handling, but I think I'm missing why this doesn't work:

```suggestion
      args.named.to_latest_kegs
```

Reply:
Ah, never mind I see. It returns an array instead of a hash.

Personally, I think it makes more sense to add `.values` to the end of the `--HEAD` handling and just deal with the array in both cases. I think it's easier to read that way and feels more natural to me. Thoughts?

Reply:
If I added `values`, there'd be no point of converting the arrays to hashes because I'd be converting them back to arrays. There'd be no way to to tell if one of the named arguments didn't have a corresponding `head` keg. If you don't think it's a big deal to print a warning if `--HEAD` is given and one of the named arguments doesn't have a corresponding `head` keg, then the code could be made simpler.

Reply:
For instance, say I have `stable` kegs for `f1` and `f2` installed, but I only have a `head` keg installed for `f1`. Assuming both were unlinked `brew link --HEAD` would start with this array of kegs: `[f1stable, f1head, f2stable]`. If I was just using the array, that'd get filtered to `[f1head]`. When traversing this array, I'd have no way of determining that there isn't a `head` keg for `f2`. The `f1` `head` keg would get linked, but nothing else would happen and there would be no `f2`-related output.

If I first convert this initial array to a hash grouped by named, I get `{f1: [f1stable, f1head], f2: [f2stable]}`.  Then, I transform the values and end up with `{f1: f1head, f2: nil}` and I'm able to easily print a warning to alert the user that `f2` doesn't have a `head` keg installed.

Sorry if I over-explained here. Just want to get on the same page so that we can figure out if either of us is missing something.

Reply:
I'm suggestion you do this which I think eliminates the first issue you raise (it's possible I'm still wrong, I haven't tested this and I won't be able to until later today):


```ruby
if args.HEAD?
  args.named.to_kegs.group_by(&:name).transform_values do |v|
    v.find do |keg|
      keg.version.head?
    end
  end.values
else
  args.named.to_latest_kegs
end
```

The warning when there's no `HEAD` keg could be moved into that `transform_values` block.

I think it's a bit cleaner this way but if you're really opposed (or if there's still something I'm missing) I won't push it

Reply:
I hadn't considered moving the warning into the `transform_values` block. I agree that it is cleaner. My new commit makes this change. The code I committed feels a bit inelegant to me. Will you let me know if there's a more idiomatic way to express what I wrote?

Reply:
`typecheck` is also getting mad at my changes, though as far as I can tell, the `reject(&:blank)` should keep any `nil` values out of the `kegs` array.

Reply:
See below: `Array#compact` will do that for you